File: app/src/main/java/com/friendlypos/Recibos/adapters/RecibosClientesAdapter.java
package com.friendlypos.Recibos.adapters;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.friendlypos.R;
import com.friendlypos.Recibos.activity.RecibosActivity;
import com.friendlypos.Recibos.modelo.receipts;
import com.friendlypos.Recibos.modelo.recibos;
import com.friendlypos.application.util.Functions;
import com.friendlypos.login.modelo.Usuarios;
import com.friendlypos.login.util.SessionPrefes;
import com.friendlypos.preventas.modelo.Numeracion;
import com.friendlypos.principal.activity.MenuPrincipal;
import com.friendlypos.principal.modelo.Clientes;

import java.util.List;

import io.realm.Realm;
import io.realm.RealmResults;
import io.realm.Sort;

/**
 * Created by DelvoM on 04/09/2018.
 */

public class RecibosClientesAdapter extends RecyclerView.Adapter<RecibosClientesAdapter.CharacterViewHolder> {

    public List<recibos> contentList;
    private RecibosActivity activity;
    private int selected_position = -1;
    private static Context QuickContext = null;
    String facturaID, clienteID;
    int tabCliente;
    String usuer;
    SessionPrefes session;
    String idUsuario;
    int nextId;
    String numFactura;

    public RecibosClientesAdapter(Context context, RecibosActivity activity, List<recibos> contentList) {
        this.contentList = contentList;
        this.activity = activity;
        this.QuickContext = context;
        session = new SessionPrefes(context);
    }

    @Override
    public RecibosClientesAdapter.CharacterViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.lista_recibos_clientes, parent, false);

        return new RecibosClientesAdapter.CharacterViewHolder(view);
    }


    @Override
    public void onBindViewHolder(final RecibosClientesAdapter.CharacterViewHolder holder, final int position) {
        final recibos recibo = contentList.get(position);
        activity.cleanTotalizeFinal();
        Realm realm = Realm.getDefaultInstance();

        String clienteId = activity.getClienteIdRecibos();


        Clientes clientes = realm.where(Clientes.class).equalTo("id", recibo.getCustomer_id()).findFirst();
        // final invoice invoice = realm.where(invoice.class).equalTo("id", recibo.getInvoice_id()).findFirst();

        final String cardCliente = clientes.getCard();
        String companyCliente = clientes.getCompanyName();
        String fantasyCliente = clientes.getFantasyName();
        String numeracionFactura = recibo.getNumeration();
        double tot = 0.0;
        RealmResults<recibos> result1 = realm.where(recibos.class).equalTo("customer_id", recibo.getCustomer_id()).findAllSorted("date", Sort.DESCENDING);

        int cant = result1.size();
        Log.d("RECIBOSCLIENTE", cant+ "");

        for(int i = 0; i<cant; i++)
        {
            int abonado1 =  result1.get(i).getAbonado();
            double total1 =  result1.get(i).getTotal();
            double pago1 =  result1.get(i).getPaid();
            double totalPagado1 = total1 - pago1;

            Log.d("PAGOSFOR2", totalPagado1 + "   " + abonado1 + "");
            activity.setTotalizarFinalCliente(totalPagado1);

            // double porPagar = recibo.getPorPagar();
            int abonado = recibo.getAbonado();
            double total = recibo.getTotal();
            double pago = recibo.getPaid();
            double totalPagado = total - pago;
            Log.d("PAGOSFOR1", totalPagado + "   " + abonado + "");

        }

        tot = activity.getTotalizarFinalCliente();
        Log.d("PAGOSFORTOT", tot + "");

        if(tot == 0.0){
            holder.cardView.setVisibility(View.GONE);
            holder.cardView.getLayoutParams().height = 0;
            ViewGroup.MarginLayoutParams layoutParams =
                    (ViewGroup.MarginLayoutParams) holder.cardView.getLayoutParams();
            layoutParams.setMargins(0, 0,0, 0);
            holder.cardView.requestLayout();
            Log.d("inactivo", "inactivo");

        }
        else{
            holder.txt_cliente_factura_card.setText(cardCliente);
            holder.txt_cliente_factura_fantasyname.setText(fantasyCliente);
            holder.txt_cliente_factura_companyname.setText(companyCliente);
        }



        holder.cardView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                int tabCliente1 = activity.getSelecClienteTabRecibos();
                if (tabCliente1 == 1) {

                    AlertDialog dialogReturnSale = new AlertDialog.Builder(activity)
                            .setTitle("Salir")
                            .setMessage("¿Desea cancelar la factura en proceso y empezar otra?")
                            .setPositiveButton("OK", new DialogInterface.OnClickListener() {

                                @Override
                                public void onClick(DialogInterface dialog, int which) {

                                    final Realm realm2 = Realm.getDefaultInstance();

                                    realm2.executeTransaction(new Realm.Transaction() {

                                        @Override
                                        public void execute(Realm realm) {

                                            Number numero = realm.where(Numeracion.class).equalTo("sale_type", "4").max("number");

                                            if (numero == null) {
                                                nextId = 1;
                                            } else {
                                                nextId = numero.intValue() - 1;
                                            }

                                        }
                                    });
                                    final Realm realm5 = Realm.getDefaultInstance();
                                    realm5.executeTransaction(new Realm.Transaction() {
                                        @Override
                                        public void execute(Realm realm5) {
                                            Numeracion numNuevo = new Numeracion(); // unmanaged
                                            numNuevo.setSale_type("4");
                                            numNuevo.setNumeracion_numero(nextId);

                                            realm5.insertOrUpdate(numNuevo);
                                            Log.d("RecNumNuevaAtras", numNuevo + "");


                                        }

                                    });

                                    int id = nextId + 1;
                                    final String idRecipiente = String.valueOf(id);
                                    final Realm realm6 = Realm.getDefaultInstance();
                                    realm6.executeTransaction(new Realm.Transaction() {
                                        @Override
                                        public void execute(Realm realm5) {

                                            RealmResults<receipts> result = realm6.where(receipts.class).equalTo("receipts_id",idRecipiente).findAll();
                                            result.deleteAllFromRealm();
                                            Log.d("ReciboBorrado", result + "");
                                        }

                                    });
                                    realm5.close();
                                    realm6.close();

                                    //   activa = 1;
                                    activity.cleanTotalizeFinal();
                                    final ProgressDialog progresRing = ProgressDialog.show(QuickContext, "Cargando", "Seleccionando Cliente", true);
                                    progresRing.setCancelable(true);
                                    new Thread(new Runnable() {
                                        @Override
                                        public void run() {
                                            try {
                                                Thread.sleep(5000);
                                            } catch (Exception e) {

                                            }
                                            progresRing.dismiss();
                                        }
                                    }).start();

                                    int pos = position;
                                    if (pos == RecyclerView.NO_POSITION) return;

                                    notifyItemChanged(selected_position);
                                    selected_position = position;
                                    notifyItemChanged(selected_position);
                                    recibos clickedDataItem = contentList.get(pos);

                                    facturaID = clickedDataItem.getInvoice_id();
                                    clienteID = clickedDataItem.getCustomer_id();
                                    double totalP = clickedDataItem.getPorPagar();

                                    Log.d("totalP", totalP + "");

                                    //  Toast.makeText(activity, facturaID + " " + clienteID + " ", Toast.LENGTH_LONG).show();

                                    tabCliente = 1;
                                    activity.setSelecClienteTabRecibos(tabCliente);
                                    activity.setClienteIdRecibos(clienteID);
                                    crearRecibo();
                                }
                            }).setNegativeButton("Cancel", new DialogInterface.OnClickListener() {

                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    dialog.cancel();
                                }
                            }).create();
                    dialogReturnSale.show();

                }else{
                    //   activa = 1;
                    activity.cleanTotalizeFinal();
                    final ProgressDialog progresRing = ProgressDialog.show(QuickContext, "Cargando", "Seleccionando Cliente", true);
                    progresRing.setCancelable(true);
                    new Thread(new Runnable() {
                        @Override
                        public void run() {
                            try {
                                Thread.sleep(5000);
                            } catch (Exception e) {

                            }
                            progresRing.dismiss();
                        }
                    }).start();

                    int pos = position;
                    if (pos == RecyclerView.NO_POSITION) return;

                    notifyItemChanged(selected_position);
                    selected_position = position;
                    notifyItemChanged(selected_position);
                    recibos clickedDataItem = contentList.get(pos);

                    facturaID = clickedDataItem.getInvoice_id();
                    clienteID = clickedDataItem.getCustomer_id();
                    double totalP = clickedDataItem.getPorPagar();

                    Log.d("totalP", totalP + "");

                    //  Toast.makeText(activity, facturaID + " " + clienteID + " ", Toast.LENGTH_LONG).show();

                    tabCliente = 1;
                    activity.setSelecClienteTabRecibos(tabCliente);
                    activity.setClienteIdRecibos(clienteID);
                    crearRecibo();
                }
            }
        });
        if(selected_position==position){
            holder.cardView.setBackgroundColor(Color.parseColor("#607d8b"));
        }
        else
        {
            holder.cardView.setBackgroundColor(Color.parseColor("#009688"));
        }

    }

    @Override
    public int getItemCount() {
        return contentList.size();
    }

    public class CharacterViewHolder extends RecyclerView.ViewHolder {

        private TextView txt_cliente_factura_card, txt_cliente_factura_fantasyname, txt_cliente_factura_companyname;
        public CardView cardView;


        public CharacterViewHolder(View view) {
            super(view);
            cardView = (CardView) view.findViewById(R.id.cardViewRecibosClientes);
            txt_cliente_factura_card = (TextView) view.findViewById(R.id.txt_cliente_factura_cardRecibos);
            txt_cliente_factura_fantasyname = (TextView) view.findViewById(R.id.txt_cliente_factura_fantasynameRecibos);
            txt_cliente_factura_companyname = (TextView) view.findViewById(R.id.txt_cliente_factura_companynameRecibos);

        }
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }

    public void crearRecibo(){

        Realm realm = Realm.getDefaultInstance();
        usuer = session.getUsuarioPrefs();
        Usuarios usuarios = realm.where(Usuarios.class).equalTo("email", usuer).findFirst();
        idUsuario = usuarios.getId();
        realm.close();

        final Realm realm2 = Realm.getDefaultInstance();

        realm2.executeTransaction(new Realm.Transaction() {

            @Override
            public void execute(Realm realm) {
                // TODO NUMERACION = VENTA DIRECTA :01, PREVENTA:02, DISTRIBUCION:03 Y PROFORMA: 04

                Number numero = realm.where(Numeracion.class).equalTo("sale_type", "4").max("number");

                if (numero == null) {
                    nextId = 1;
                } else {
                    nextId = numero.intValue() + 1;
                }
                int valor = numero.intValue();

                int length = String.valueOf(valor).length();

                if(length == 1){
                    numFactura = idUsuario + "04-" +  "000000" + nextId;
                }
                else if(length == 2){
                    numFactura = idUsuario + "04-" +  "00000" + nextId;
                }
                else if(length == 3){
                    numFactura = idUsuario + "04-" +  "0000" + nextId;
                }
                else if(length == 4){
                    numFactura = idUsuario + "04-" +  "000" + nextId;
                }
                else if(length == 5){
                    numFactura = idUsuario + "04-" +  "00" + nextId;
                }
                else if(length == 6){
                    numFactura = idUsuario + "04-" +  "0" + nextId;
                }
                else if(length == 7){
                    numFactura = idUsuario + "04-" +  nextId;
                }
            }
        });


        final Realm realmRecibo = Realm.getDefaultInstance();
        realmRecibo.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realmRecibo) {
                receipts receipt = new receipts(); // unmanaged

                receipt.setReceipts_id(String.valueOf(nextId));
                receipt.setCustomer_id(clienteID);
                receipt.setCustomer_id(clienteID);
                receipt.setReference(numFactura);
                receipt.setDate(Functions.getDate());

                realmRecibo.insertOrUpdate(receipt);
                Log.d("ReciboNuevo", receipt + "");
                activity.setReceipts_id_num(String.valueOf(nextId));


            }

        });
        realmRecibo.close();


        final Realm realm5 = Realm.getDefaultInstance();
        realm5.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm5) {
                Numeracion numNuevo = new Numeracion(); // unmanaged
                numNuevo.setSale_type("4");
                numNuevo.setNumeracion_numero(nextId);
                realm5.insertOrUpdate(numNuevo);
                Log.d("NumRecibosNueva", numNuevo + "");


            }

        });
        realm5.close();


    }

}

File: app/src/main/java/com/friendlypos/Recibos/adapters/RecibosClientesAdapter.kt
package com.friendlysystemgroup.friendlypos.Recibos.adapters

import android.app.ProgressDialog
import android.content.Context
import android.graphics.Color
import android.text.SpannableString
import android.text.Spanned
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.RecyclerView
import com.friendlysystemgroup.friendlypos.Recibos.activity.RecibosActivity
import com.friendlysystemgroup.friendlypos.Recibos.modelo.receipts
import com.friendlysystemgroup.friendlypos.Recibos.modelo.recibos
import com.friendlysystemgroup.friendlypos.application.util.Functions.date
import com.friendlysystemgroup.friendlypos.databinding.ListaRecibosClientesBinding
import com.friendlysystemgroup.friendlypos.login.modelo.Usuarios
import com.friendlysystemgroup.friendlypos.login.util.SessionPrefes
import com.friendlysystemgroup.friendlypos.preventas.modelo.Numeracion
import com.friendlysystemgroup.friendlypos.principal.modelo.Clientes
import io.realm.Realm
import io.realm.RealmResults
import io.realm.Sort
import uk.co.chrisjenx.calligraphy.CalligraphyTypefaceSpan
import uk.co.chrisjenx.calligraphy.TypefaceUtils

/**
 * Created by DelvoM on 04/09/2018.
 */
class RecibosClientesAdapter(
    private val context: Context,
    private val activity: RecibosActivity,
    var contentList: List<recibos>
) : RecyclerView.Adapter<RecibosClientesAdapter.CharacterViewHolder>() {
    private var selected_position = -1
    var facturaID: String? = null
    var clienteID: String? = null
    var tabCliente: Int = 0
    private var usuer: String? = null
    private val session: SessionPrefes = SessionPrefes(context)
    private var idUsuario: String? = null
    private var nextId: Int = 0
    private var numFactura: String? = null

    init {
        QuickContext = context
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CharacterViewHolder {
        val binding = ListaRecibosClientesBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return CharacterViewHolder(binding)
    }

    override fun onBindViewHolder(holder: CharacterViewHolder, position: Int) {
        val recibo = contentList[position]
        activity.cleanTotalizeFinal()
        val realm = Realm.getDefaultInstance()

        val clientes = realm.where(Clientes::class.java)
            .equalTo("id", recibo.customer_id)
            .findFirst()

        val result1: RealmResults<recibos> = realm.where(recibos::class.java)
            .equalTo("customer_id", recibo.customer_id)
            .sort("date", Sort.DESCENDING)
            .findAll()

        val cant = result1.size
        Log.d("RECIBOSCLIENTE", cant.toString())

        for (i in 0 until cant) {
            result1[i]?.let { abonado1 ->
                val total1 = abonado1.total
                val pago1 = abonado1.paid
                val totalPagado1 = total1 - pago1

                Log.d("PAGOSFOR2", "$totalPagado1   ${abonado1.abonado}")
                activity.setTotalizarFinalCliente(totalPagado1)
            }
        }

        val tot = activity.getTotalizarFinalCliente()
        Log.d("PAGOSFORTOT", tot.toString())

        if (tot == 0.0) {
            holder.binding.cardViewRecibosClientes.visibility = View.GONE
            holder.binding.cardViewRecibosClientes.layoutParams.height = 0
            val layoutParams = holder.binding.cardViewRecibosClientes.layoutParams as ViewGroup.MarginLayoutParams
            layoutParams.setMargins(0, 0, 0, 0)
            holder.binding.cardViewRecibosClientes.requestLayout()
            Log.d("inactivo", "inactivo")
        } else {
            clientes?.let { cliente ->
                holder.binding.txtClienteFacturaCardRecibos.text = cliente.card
                holder.binding.txtClienteFacturaFantasynameRecibos.text = cliente.fantasyName
                holder.binding.txtClienteFacturaCompanynameRecibos.text = cliente.companyName
            }
        }

        holder.binding.cardViewRecibosClientes.setOnClickListener {
            val tabCliente1 = activity.selecClienteTabRecibos
            if (tabCliente1 == 1) {
                AlertDialog.Builder(activity)
                    .setTitle("Salir")
                    .setMessage("¿Desea cancelar la factura en proceso y empezar otra?")
                    .setPositiveButton("OK") { _, _ ->
                        val realm2 = Realm.getDefaultInstance()
                        realm2.executeTransaction { r ->
                            val numero = r.where(Numeracion::class.java)
                                .equalTo("sale_type", "4").max("number")
                            nextId = numero?.toInt()?.minus(1) ?: 1
                        }
                        
                        val realm5 = Realm.getDefaultInstance()
                        realm5.executeTransaction { r ->
                            val numNuevo = Numeracion()
                            numNuevo.sale_type = "4"
                            numNuevo.numeracion_numero = nextId

                            r.insertOrUpdate(numNuevo)
                            Log.d("RecNumNuevaAtras", numNuevo.toString())
                        }

                        val id = nextId + 1
                        val idRecipiente = id.toString()
                        val realm6 = Realm.getDefaultInstance()
                        realm6.executeTransaction { r ->
                            val result = r.where(receipts::class.java)
                                .equalTo("receipts_id", idRecipiente)
                                .findAll()
                            result.deleteAllFromRealm()
                            Log.d("ReciboBorrado", result.toString())
                        }
                        realm5.close()
                        realm6.close()

                        activity.cleanTotalizeFinal()
                        val progresRing = ProgressDialog.show(
                            QuickContext,
                            "Cargando",
                            "Seleccionando Cliente",
                            true
                        )
                        progresRing.setCancelable(true)
                        Thread {
                            try {
                                Thread.sleep(5000)
                            } catch (e: Exception) {
                                // No action needed
                            }
                            progresRing.dismiss()
                        }.start()

                        val pos = holder.adapterPosition
                        if (pos == RecyclerView.NO_POSITION) return@setPositiveButton

                        notifyItemChanged(selected_position)
                        selected_position = pos
                        notifyItemChanged(selected_position)
                        val clickedDataItem = contentList[pos]

                        facturaID = clickedDataItem.invoice_id
                        clienteID = clickedDataItem.customer_id
                        Log.d("totalP", clickedDataItem.porPagar.toString())

                        tabCliente = 1
                        activity.selecClienteTabRecibos = tabCliente
                        activity.clienteIdRecibos = clienteID
                        crearRecibo()
                    }
                    .setNegativeButton("Cancel") { dialog, _ -> dialog.cancel() }
                    .create()
                    .show()
            } else {
                activity.cleanTotalizeFinal()

                // Usar ProgressDialog con estilos personalizados
                val progresRing = ProgressDialog(QuickContext)
                val message = "Seleccionando Cliente"
                val titulo = "Cargando"
                val spannableString = SpannableString(message)
                val spannableStringTitulo = SpannableString(titulo)

                val typefaceSpan = CalligraphyTypefaceSpan(
                    TypefaceUtils.load(
                        QuickContext?.assets, "font/monse.otf"
                    )
                )
                spannableString.setSpan(
                    typefaceSpan,
                    0,
                    message.length,
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                )
                spannableStringTitulo.setSpan(typefaceSpan, 0, titulo.length, Spanned.SPAN_PRIORITY)

                progresRing.setTitle(spannableStringTitulo)
                progresRing.setMessage(spannableString)
                progresRing.setProgressStyle(ProgressDialog.STYLE_SPINNER)
                progresRing.isIndeterminate = true
                progresRing.setCancelable(true)
                progresRing.show()
                
                Thread {
                    try {
                        Thread.sleep(5000)
                    } catch (e: Exception) {
                        // No action needed
                    }
                    progresRing.dismiss()
                }.start()

                val pos = holder.adapterPosition
                if (pos == RecyclerView.NO_POSITION) return@setOnClickListener

                notifyItemChanged(selected_position)
                selected_position = pos
                notifyItemChanged(selected_position)
                val clickedDataItem = contentList[pos]

                facturaID = clickedDataItem.invoice_id
                clienteID = clickedDataItem.customer_id
                Log.d("totalP", clickedDataItem.porPagar.toString())

                tabCliente = 1
                activity.selecClienteTabRecibos = tabCliente
                activity.clienteIdRecibos = clienteID
                crearRecibo()
            }
        }
        
        holder.binding.cardViewRecibosClientes.setBackgroundColor(
            if (selected_position == position) Color.parseColor("#d1d3d4")
            else Color.parseColor("#FFFFFF")
        )
    }

    override fun getItemCount(): Int = contentList.size

    inner class CharacterViewHolder(val binding: ListaRecibosClientesBinding) : 
        RecyclerView.ViewHolder(binding.root)

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
    }

    private fun crearRecibo() {
        val realm = Realm.getDefaultInstance()
        usuer = session.usuarioPrefs
        val usuarios = realm.where(Usuarios::class.java)
            .equalTo("email", usuer)
            .findFirst()
        idUsuario = usuarios?.id
        realm.close()

        val realm2 = Realm.getDefaultInstance()
        realm2.executeTransaction { r ->
            val numero = r.where(Numeracion::class.java)
                .equalTo("sale_type", "4")
                .max("number")

            nextId = numero?.toInt()?.plus(1) ?: 1
            val valor = numero?.toInt() ?: 0

            numFactura = when (valor.toString().length) {
                1 -> "${idUsuario}04-000000$nextId"
                2 -> "${idUsuario}04-00000$nextId"
                3 -> "${idUsuario}04-0000$nextId"
                4 -> "${idUsuario}04-000$nextId"
                5 -> "${idUsuario}04-00$nextId"
                6 -> "${idUsuario}04-0$nextId"
                else -> "${idUsuario}04-$nextId"
            }
        }

        val realmRecibo = Realm.getDefaultInstance()
        realmRecibo.executeTransaction { r ->
            val receipt = receipts()
            receipt.receipts_id = nextId.toString()
            receipt.customer_id = clienteID
            receipt.reference = numFactura
            receipt.date = date

            r.insertOrUpdate(receipt)
            Log.d("ReciboNuevo", receipt.toString())
            activity.receipts_id_num = nextId.toString()
        }
        realmRecibo.close()

        val realm5 = Realm.getDefaultInstance()
        realm5.executeTransaction { r ->
            val numNuevo = Numeracion()
            numNuevo.sale_type = "4"
            numNuevo.numeracion_numero = nextId
            r.insertOrUpdate(numNuevo)
            Log.d("NumRecibosNueva", numNuevo.toString())
        }
        realm5.close()
    }

    companion object {
        private var QuickContext: Context? = null
    }
}

